### DISCUSSION TITLE
cannot serve when pb_data is a symlink to NAS on MacOS

### DISCUSSION THREAD

**Discussion by wiezmankimchi** [pocketbase/pocketbase#5960] [Q&A]
When running `pocketbase serve` where `pb_data` is a symlink to a folder on NAS , error is produced:

`user:pocketbase/ $ ./pocketbase serve                                                                               [20:51:14]
libc_darwin.go:197:Xsrandomdev: TODOTODO 
goroutine 1 [running]:
runtime/debug.Stack()
	/opt/hostedtoolcache/go/1.23.3/x64/src/runtime/debug/stack.go:26 +0x64
modernc.org/libc.todo({0x0?, 0x10?}, {0x0?, 0x0?, 0x0?})
	/home/runner/go/pkg/mod/modernc.org/libc@v1.61.2/etc.go:102 +0x108
modernc.org/libc.Xsrandomdev(...)
	/home/runner/go/pkg/mod/modernc.org/libc@v1.61.2/libc_darwin.go:197
modernc.org/sqlite/lib._fillInUnixFile(0x14000b04d80, 0x103218160, 0x9, 0x10ab00db0, 0x10ab00f3c, 0x0)
	/home/runner/go/pkg/mod/modernc.org/sqlite@v1.34.1/lib/sqlite_darwin_arm64.go:31368 +0x390
modernc.org/sqlite/lib._unixOpen(0x14000b04d80, 0x103218160, 0x10ab00f3c, 0x10ab00db0, 0x106, 0x10a50010c)
	/home/runner/go/pkg/mod/modernc.org/sqlite@v1.34.1/lib/sqlite_darwin_arm64.go:31901 +0x8b0
modernc.org/sqlite/lib._sqlite3OsOpen(0x14000b04d80?, 0x10ab00fd1?, 0x10a90016c?, 0x100000201?, 0xac00028?, 0x101756f7c?)
	/home/runner/go/pkg/mod/modernc.org/sqlite@v1.34.1/lib/sqlite_darwin_arm64.go:17433 +0x40
modernc.org/sqlite/lib._sqlite3PagerOpen(0x14000b04d80, 0x103218160, 0x10a400628, 0x10a90016c, 0x88, 0x0, 0x106, 0x102757cd8)
	/home/runner/go/pkg/mod/modernc.org/sqlite@v1.34.1/lib/sqlite_darwin_arm64.go:42804 +0x59c
modernc.org/sqlite/lib._sqlite3BtreeOpen(0x14000b04d80, 0x103218160, 0x10a90016c, 0x10ab00828, 0x10ab00ae8, 0x0, 0x106)
	/home/runner/go/pkg/mod/modernc.org/sqlite@v1.34.1/lib/sqlite_darwin_arm64.go:52863 +0x27c
modernc.org/sqlite/lib._openDatabase(0x14000b04d80, 0x10a600030, 0x10a600020, 0x10046, 0x0)
	/home/runner/go/pkg/mod/modernc.org/sqlite@v1.34.1/lib/sqlite_darwin_arm64.go:168934 +0x5d4
modernc.org/sqlite/lib.Xsqlite3_open_v2(...)
	/home/runner/go/pkg/mod/modernc.org/sqlite@v1.34.1/lib/sqlite_darwin_arm64.go:169029
modernc.org/sqlite.(*conn).openV2(0x14000aa7200, {0x140005ce0e0, 0xf}, {0x0, 0x0}, 0x10046)
	/home/runner/go/pkg/mod/modernc.org/sqlite@v1.34.1/sqlite.go:1369 +0xe4
modernc.org/sqlite.newConn({0x140005ce0e0, 0xd6})
	/home/runner/go/pkg/mod/modernc.org/sqlite@v1.34.1/sqlite.go:839 +0x168
modernc.org/sqlite.(*Driver).Open(0x1033f1960, {0x140005ce0e0, 0xd6})
	/home/runner/go/pkg/mod/modernc.org/sqlite@v1.34.1/sqlite.go:1965 +0x34
database/sql.dsnConnector.Connect(...)
	/opt/hostedtoolcache/go/1.23.3/x64/src/database/sql/sql.go:801
database/sql.(*DB).conn(0x140006fa820, {0x1027742e8, 0x10341b3a0}, 0x1)
	/opt/hostedtoolcache/go/1.23.3/x64/src/database/sql/sql.go:1423 +0x98c
database/sql.(*DB).exec(0x140006fa820, {0x1027742e8, 0x10341b3a0}, {0x140006e77a0, 0x6b}, {0x0, 0x0, 0x0}, 0x98?)
	/opt/hostedtoolcache/go/1.23.3/x64/src/database/sql/sql.go:1681 +0x40
database/sql.(*DB).ExecContext.func1(0x6b?)
	/opt/hostedtoolcache/go/1.23.3/x64/src/database/sql/sql.go:1664 +0x54
database/sql.(*DB).retry(0x10?, 0x14000b116f0)
	/opt/hostedtoolcache/go/1.23.3/x64/src/database/sql/sql.go:1568 +0x4c
database/sql.(*DB).ExecContext(0x14000b11788?, {0x1027742e8?, 0x10341b3a0?}, {0x140006e77a0?, 0xf90aa3800b117e8?}, {0x0?, 0x14000b11798?, 0x1016ad194?})
	/opt/hostedtoolcache/go/1.23.3/x64/src/database/sql/sql.go:1663 +0x94
database/sql.(*DB).Exec(0x10a1fead8?, {0x140006e77a0?, 0x2?}, {0x0?, 0x18?, 0x18?})
	/opt/hostedtoolcache/go/1.23.3/x64/src/database/sql/sql.go:1677 +0x48
github.com/pocketbase/dbx.(*Query).execute(0x14000b04cc0)
	/home/runner/go/pkg/mod/github.com/pocketbase/dbx@v1.10.1/query.go:231 +0x184
github.com/pocketbase/dbx.(*Query).Execute.func1()
	/home/runner/go/pkg/mod/github.com/pocketbase/dbx@v1.10.1/query.go:207 +0x28
github.com/pocketbase/dbx.(*Query).execWrap(0x10?, 0x140006e7650?)
	/home/runner/go/pkg/mod/github.com/pocketbase/dbx@v1.10.1/query.go:366 +0x40
github.com/pocketbase/dbx.(*Query).Execute(0x14000b04cc0)
	/home/runner/go/pkg/mod/github.com/pocketbase/dbx@v1.10.1/query.go:205 +0x88
github.com/pocketbase/pocketbase/core.(*MigrationsRunner).initMigrationsTable(0x14000aa7180)
	/home/runner/work/pocketbase/pocketbase/core/migrations_runner.go:260 +0xa0
github.com/pocketbase/pocketbase/core.(*MigrationsRunner).Up(0x14000aa7180)
	/home/runner/work/pocketbase/pocketbase/core/migrations_runner.go:127 +0x20
github.com/pocketbase/pocketbase/core.(*BaseApp).RunSystemMigrations(...)
	/home/runner/work/pocketbase/pocketbase/core/base.go:708
github.com/pocketbase/pocketbase/core.(*BaseApp).Bootstrap.func1(0x1400078cea0?)
	/home/runner/work/pocketbase/pocketbase/core/base.go:405 +0x100
github.com/pocketbase/pocketbase/tools/hook.(*Hook[...]).Trigger.func1()
	/home/runner/work/pocketbase/pocketbase/tools/hook/hook.go:167 +0x6c
github.com/pocketbase/pocketbase/tools/hook.(*Event).Next(...)
	/home/runner/work/pocketbase/pocketbase/tools/hook/event.go:32
github.com/pocketbase/pocketbase/core.(*BaseApp).registerCollectionHooks.func15(0x14000aaa798)
	/home/runner/work/pocketbase/pocketbase/core/collection_model.go:295 +0x38
github.com/pocketbase/pocketbase/tools/hook.(*Hook[...]).Trigger.func1()
	/home/runner/work/pocketbase/pocketbase/tools/hook/hook.go:167 +0x6c
github.com/pocketbase/pocketbase/tools/hook.(*Event).Next(...)
	/home/runner/work/pocketbase/pocketbase/tools/hook/event.go:32
github.com/pocketbase/pocketbase/plugins/jsvm.Register.func1(0x10?)
	/home/runner/work/pocketbase/pocketbase/plugins/jsvm/jsvm.go:131 +0x3c
github.com/pocketbase/pocketbase/tools/hook.(*Hook[...]).Trigger.func1()
	/home/runner/work/pocketbase/pocketbase/tools/hook/hook.go:167 +0x6c
github.com/pocketbase/pocketbase/tools/hook.(*Event).Next(...)
	/home/runner/work/pocketbase/pocketbase/tools/hook/event.go:32
github.com/pocketbase/pocketbase/core.(*BaseApp).registerAutobackupHooks.func2(0x0?)
	/home/runner/work/pocketbase/pocketbase/core/base_backup.go:332 +0x40
github.com/pocketbase/pocketbase/tools/hook.(*Hook[...]).Trigger.func1()
	/home/runner/work/pocketbase/pocketbase/tools/hook/hook.go:167 +0x6c
github.com/pocketbase/pocketbase/tools/hook.(*Event).Next(...)
	/home/runner/work/pocketbase/pocketbase/tools/hook/event.go:32
github.com/pocketbase/pocketbase/tools/hook.(*Hook[...]).Trigger(0x102783980, 0x14000aaa798, {0x14000b11d30, 0x1, 0x140001b8030})
	/home/runner/work/pocketbase/pocketbase/tools/hook/hook.go:171 +0x3c4
github.com/pocketbase/pocketbase/core.(*BaseApp).Bootstrap(0x140005eb188)
	/home/runner/work/pocketbase/pocketbase/core/base.go:382 +0xec
github.com/pocketbase/pocketbase.(*PocketBase).Execute(0x140001f11d0)
	/home/runner/work/pocketbase/pocketbase/pocketbase.go:160 +0x38
github.com/pocketbase/pocketbase.(*PocketBase).Start(0x140001f11d0)
	/home/runner/work/pocketbase/pocketbase/pocketbase.go:150 +0x84
main.main()
	/home/runner/work/pocketbase/pocketbase/examples/base/main.go:119 +0x4bc`

Working folder content is (` /Volumes/docker/pocketbase/v23/` is located on NAS)
`total 81216
    0 lrwxr-xr-x  1 user  staff        38 Nov 26 20:49 pb_data -> /Volumes/docker/pocketbase/v23/pb_data
    0 lrwxr-xr-x  1 user  staff        39 Nov 26 20:49 pb_hooks -> /Volumes/docker/pocketbase/v23/pb_hooks
    0 lrwxr-xr-x  1 user  staff        44 Nov 26 20:50 pb_migrations -> /Volumes/docker/pocketbase/v23/pb_migrations
    0 lrwxr-xr-x  1 user  staff        40 Nov 26 20:50 pb_public -> /Volumes/docker/pocketbase/v23/pb_public
81216 -rwxr-xr-x@ 1 user  staff  41582146 Nov 25 01:39 pocketbase`

pocketbase version: 0.23.1
`user:pocketbase/ $ ./pocketbase --version                                                                           [20:54:26]
pocketbase version 0.23.1`

**Comment by wiezmankimchi** [comment]
The same error appears when trying to run `./pocketbase --dir /Volumes/docker/pocketbase/v23/pb_data serve  `

**Comment by ganigeorgiev** [comment]
Unfortunately the stack trace doesn't really help me narrowing the issue so I don't think I'll be able to help you (I don't have access to macOS, nor the appropriate env setup to trigger the error).

I'll move the issue to a Q&A discussion, maybe somebody else would be able to assist you.

I'll leave couple things that you could check:

0. _For accessing SQLite over a network filesystem you may want to check the caveats mentioned in https://www.sqlite.org/useovernet.html._

1. Make sure that you are using the correct PocketBase executable for your platform.

2. To rule out a PocketBase issue, you can try building and running the following minimal Go program with both the CGO and pure Go drivers:
    ```go
    package main

    import (
      "database/sql"
      "log"

      _ "modernc.org/sqlite"

      // to ensure that the problem happens only with the pure Go driver
      // experiment commenting the above driver and uncomment the below one
      // while building with CGO_ENABLED=1
      // (the driver name in the Open call should be "sqlite3")
      // _ "github.com/mattn/go-sqlite3"
    )

    func main() {
      db, err := sql.Open("sqlite", "/Volumes/docker/pocketbase/v23/pb_data/data.db")
      if err != nil {
        log.Fatal(err)
      }
      defer db.Close()

      err = db.Ping()
      if err != nil {
        log.Fatal(err)
      }

      log.Println("Connected successfully!")
    }
    ```
    If it doesn't work only with the pure Go driver then you can try reporting it in [modernc.org/sqlite](https://gitlab.com/cznic/sqlite).

    To build PocketBase with different drivers you can check https://pocketbase.io/docs/go-overview/#custom-sqlite-driver.
